cmd/authd/daemon/migration_test.go

```diff
@@ -0,0 +1,109 @@
+package daemon
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestMigrateOldCacheDir(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		oldDirExists     bool
+		newDirExists     bool
+		oldDirUnreadable bool
+		newDirUnreadable bool
+
+		wantErr          bool
+		wantOldDirExists bool
+		wantNewDirExists bool
+	}{
+		"Success if old dir does not exist": {oldDirExists: false, newDirExists: false},
+		"Success if old dir exists and new dir does not": {
+			oldDirExists:     true,
+			newDirExists:     false,
+			wantOldDirExists: false,
+			wantNewDirExists: true,
+		},
+		"Success if old dir exists and new dir exists": {
+			oldDirExists:     true,
+			newDirExists:     true,
+			wantOldDirExists: true,
+			wantNewDirExists: true,
+		},
+		"Success if old dir exists but is unreadable": {
+			oldDirExists:     true,
+			oldDirUnreadable: true,
+		},
+
+		"Error if new dir exists but is unreadable": {
+			oldDirExists:     true,
+			newDirExists:     true,
+			newDirUnreadable: true,
+			wantErr:          true,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			oldParentDir := t.TempDir()
+			newParentDir := t.TempDir()
+			oldDir := filepath.Join(oldParentDir, "cache")
+			newDir := filepath.Join(newParentDir, "cache")
+
+			if tc.oldDirExists {
+				err := os.Mkdir(oldDir, 0700)
+				require.NoError(t, err)
+
+				if tc.oldDirUnreadable {
+					err := os.Chmod(oldParentDir, 0000)
+					require.NoError(t, err)
+					// Ensure that the directory is readable after the test, to avoid issues with cleanup
+					defer func() {
+						//nolint:gosec // G302 Permissions 0700 are not insecure for a directory
+						err := os.Chmod(oldParentDir, 0700)
+						require.NoError(t, err)
+					}()
+				}
+			}
+
+			if tc.newDirExists {
+				err := os.Mkdir(newDir, 0700)
+				require.NoError(t, err)
+
+				if tc.newDirUnreadable {
+					err := os.Chmod(newParentDir, 0000)
+					require.NoError(t, err)
+					// Ensure that the directory is readable after the test, to avoid issues with cleanup
+					defer func() {
+						//nolint:gosec // G302 Permissions 0700 are not insecure for a directory
+						err := os.Chmod(newParentDir, 0700)
+						require.NoError(t, err)
+					}()
+				}
+			}
+
+			err := migrateOldCacheDir(oldDir, newDir)
+			if tc.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
```

Comment:
What about using `require.ErrorContains` instead? Or you can likely use even `require.ErrorIs` here using `WantError error` and checking for the same error kind (as you're passing it with `%w`).

Reply:
This adds an expectation to `migrateOldCacheDir` (and `fileutils.FileExists`) that it returns an `os.ErrPermission` if it fails because of permission issues. In this case I'm fine with that, because it's a reasonable expectation and it's easy to check for this error, so I pushed a fixup commit which does that.